sentinel-core/src/main/java/com/alibaba/csp/sentinel/spi/SpiLoader.java

```diff
@@ -0,0 +1,541 @@
+
+    /**
+     * Load the first-found Provider instance
+     *
+     * @return Provider instance of first-found specific
+     */
+    public S loadFirstInstance() {
+        load();
+
+        if (classList.size() == 0) {
+            return null;
+        }
+
+        Class<? extends S> serviceClass = classList.get(0);
+        S instance = createInstance(serviceClass);
+        return instance;
+    }
+
+    /**
+     * Load the first-found Provider instance,if not found, return default Provider instance
+     *
+     * @return Provider instance
+     */
+    public S loadFirstInstanceOrDefault() {
+        load();
+
+        for (Class<? extends S> clazz : classList) {
+            if (defaultClass == null || clazz != defaultClass) {
+                return createInstance(clazz);
+            }
+        }
+
+        return loadDefaultInstance();
+    }
+
+    /**
+     * Load default Provider instance
+     * Provider class with @Spi(isDefault = true)
+     *
+     * @return default Provider instance
+     */
+    public S loadDefaultInstance() {
+        load();
+
+        if (defaultClass == null) {
+            return null;
+        }
+
+        return createInstance(defaultClass);
+    }
+
+    /**
+     * Load instance by specific class type
+     *
+     * @param clazz class type
+     * @return Provider instance
+     */
+    public S loadInstance(Class<? extends S> clazz) {
+        AssertUtil.notNull(clazz, "SPI class cannot be null");
+
+        if (clazz.equals(service)) {
+            fail(clazz.getName() + " is not subtype of " + service.getName());
+        }
+
+        load();
+
+        if (!classMap.containsValue(clazz)) {
+            fail(clazz.getName() + " is not Provider class of " + service.getName() + ",check if it is in the SPI file?");
+        }
+
+        return createInstance(clazz);
+    }
+
+    /**
+     * Load instance by aliasName of Provider class
+     *
+     * @param aliasName aliasName of Provider class
+     * @return Provider instance
+     */
+    public S loadInstance(String aliasName) {
+        AssertUtil.notEmpty(aliasName, "aliasName cannot be empty");
+
+        load();
+
+        Class<? extends S> clazz = classMap.get(aliasName);
+        if (clazz == null) {
+            fail("no Provider class's aliasName is " + aliasName);
+        }
+
+        return createInstance(clazz);
+    }
+
+    /**
+     * Reset all fields of current SpiLoader instance
+     */
+    public synchronized void reset() {
+        SPI_LOADER_MAP.remove(service.getName());
+        classList.clear();
+        sortedClassList.clear();
+        classMap.clear();
+        singletonMap.clear();
+        defaultClass = null;
+        loaded.set(false);
+    }
+
+    /**
+     * Load the Provider class from Provider configuration file
+     */
+    public void load() {
+        if (!loaded.compareAndSet(false, true)) {
+            return;
+        }
+
+        String fullFileName = SPI_FILE_PREFIX + service.getName();
```

Comment:
JDK's `ServiceProvider` is convenient to use, since we here want to enrich its functions, support our custom `Spi` annotation, so load them manually. This implementation refers to `ServerProvider`.

Reply:
> JDK's `ServiceProvider` is convenient to use, since we here want to enrich its functions, support our custom `Spi` annotation, so load them manually. This implementation refers to `ServerProvider`.

Ok i get it. It's ok then.